### ISSUE TITLE
add scf.rof

### ISSUE THREAD

**Issue by superlopuh** [xdslproject/xdsl#2039]

**Comment by alexarice** [comment]
Please elaborate. Scf doesn't have this operation

**Comment by superlopuh** [comment]
Yep, our backend dialects do, this is a relatively important optimisation in riscv since there's a `zero` register that's always zero. If you count up from 0 to `n`, you need to keep `i` and `n` in a register, but if you count down from `n` to 0, you can just compare with zero. By the time we get to `riscv_scf`, we often don't have enough semantic knowledge to tell that the loop iteration can be reverted, but tiling from linalg directly into rof loops is lega. It's a long-term nice-to have, let's keep the issue around for the time being.

**Comment by alexarice** [comment]
I think I disagree with adding operations to dialects which are not upstream. Is there a reason you can't lower straight to riscv_scf?

**Comment by superlopuh** [comment]
Yeah I broadly agree. I'm not sure what the optimal way to do this would be, other than an `xdsl_scf` dialect a bit like the pointer one. And no, because there's an abstraction gap there of a loop that iterates in a specific order in a target-independent way, we might want to use this to target x86 or something else.
